### DISCUSSION TITLE
Corruption self tests are skipped when verify_integrity is failed

### DISCUSSION THREAD

**Discussion by shuishuiwawa** [openssl/openssl#24618] [Q&A] [answered]
## Description
I built OpenSSL from source using version 3.0.1, and also installed FIPS provider. I tried to run the self test manually for FIPS provider by running `./provider_status_test -config /opt/ossl3/ssl/openssl-fips.cnf -provider_name fips`, the command itself runs successfully, but the result looks weird. See it below. When it comes to on demand failure, only one test case is run. Failure to verify integrity of the provider will cause provider load failure, therefore we cannot test any other algorithms. This is the expected behavior of failure to verify integrity, but the provider_status_test tries to do some corruption tests for some algorithms, including `verify_integrity`. What I expect is that other corruption tests will be run after module verification failure.
but the test case tries to corrupt integrity and 
```
[root@localhost test]# ./provider_status_test -config /opt/ossl3/ssl/openssl-fips.cnf -provider_name fips
# 
# On Loading
# HMAC : (Module_Integrity) : Pass
# HMAC : (Install_Integrity) : Pass
1..1
start self test...# 
# On Demand
# HMAC : (Module_Integrity) : Pass
# HMAC : (Install_Integrity) : Pass
# SHA1 : (KAT_Digest) : Pass
# SHA2 : (KAT_Digest) : Pass
# SHA3 : (KAT_Digest) : Pass
# TDES : (KAT_Cipher) : Pass
# AES_GCM : (KAT_Cipher) : Pass
# AES_ECB_Decrypt : (KAT_Cipher) : Pass
# RSA : (KAT_Signature) : RNG : (Continuous_RNG_Test) : Pass
# Pass
# ECDSA : (PCT_Signature) : Pass
# ECDSA : (PCT_Signature) : Pass
# DSA : (PCT_Signature) : Pass
# TLS13_KDF_EXTRACT : (KAT_KDF) : Pass
# TLS13_KDF_EXPAND : (KAT_KDF) : Pass
# TLS12_PRF : (KAT_KDF) : Pass
# PBKDF2 : (KAT_KDF) : Pass
# SSHKDF : (KAT_KDF) : Pass
# KBKDF : (KAT_KDF) : Pass
# HKDF : (KAT_KDF) : Pass
# SSKDF : (KAT_KDF) : Pass
# X963KDF : (KAT_KDF) : Pass
# X942KDF : (KAT_KDF) : Pass
# HASH : (DRBG) : Pass
# CTR : (DRBG) : Pass
# HMAC : (DRBG) : Pass
# DH : (KAT_KA) : Pass
# ECDH : (KAT_KA) : Pass
# RSA_Encrypt : (KAT_AsymmetricCipher) : Pass
# RSA_Decrypt : (KAT_AsymmetricCipher) : Pass
# RSA_Decrypt : (KAT_AsymmetricCipher) : Pass
# 
# On Demand Failure
# HMAC : (Module_Integrity) : Fail
ok 1 - test_provider_status
```

## Analysis
After checking the source code, I see module verification is always run in SELF_TEST_post in self_test.c. Once it fails, the program will go to end, and skip the `SELF_TEST_kats` right before `end` section. Since I see some corruption code in `self_test_digests` and some other tests, I expect other corruption tests can be triggered after the corruption for module verification.

Would you please confirm if it's a bug or expected behavior? Thank you!

**Comment by paulidale** [comment]
Did you follow the installation instructions from the security policy?  I suspect not because 3.0.1 is **not** and **never** will be FIPS validated.  Currently, only versions 3.0.8 and 3.0.9 are on the validated list.

Moreover, the installation instructions include steps which create the integrity checksum which is what is failing for you.

**Reply by shuishuiwawa** [reply]
Thanks @paulidale  for your prompt reply. I think it's not related to the version is FIPS validated or not. It's because of the code logic in self_test.c.  I am trying to run `provider_status_test` under `test` directory. There are 2 on demand self test cases, one is happy path, the other is negative. See the screenshot.
![provider_status_test](https://github.com/openssl/openssl/assets/10859404/cd079233-cd18-4a34-92a2-669f0429713e)
The happy path works as expected, but the other one does not work as my expectation (IMO). You can see the result in the issue description. What I expect is as below (after making a workaround in self_test.c). My expectation is to let the failed cases continue even though the module verification failed intentionally because of corruption test. Please correct me if I am wrong. I will also try 3.0.8 or 3.0.9 which is FIPS validated. Thanks again!
```
[root@localhost openssl]# test/provider_status_test -config /opt/ossl3/ssl/openssl-fips.cnf -provider_name fips
# 
# On Loading
# HMAC : (Module_Integrity) : Pass
# HMAC : (Install_Integrity) : Pass
1..1
# 
# On Demand
# HMAC : (Module_Integrity) : Pass
# HMAC : (Install_Integrity) : Pass
# SHA1 : (KAT_Digest) : Pass
# SHA2 : (KAT_Digest) : Pass
# SHA3 : (KAT_Digest) : Pass
# TDES : (KAT_Cipher) : Pass
# AES_GCM : (KAT_Cipher) : Pass
# AES_ECB_Decrypt : (KAT_Cipher) : Pass
# RSA : (KAT_Signature) : RNG : (Continuous_RNG_Test) : Pass
# Pass
# ECDSA : (PCT_Signature) : Pass
# ECDSA : (PCT_Signature) : Pass
# DSA : (PCT_Signature) : Pass
# TLS13_KDF_EXTRACT : (KAT_KDF) : Pass
# TLS13_KDF_EXPAND : (KAT_KDF) : Pass
# TLS12_PRF : (KAT_KDF) : Pass
# PBKDF2 : (KAT_KDF) : Pass
# SSHKDF : (KAT_KDF) : Pass
# KBKDF : (KAT_KDF) : Pass
# HKDF : (KAT_KDF) : Pass
# SSKDF : (KAT_KDF) : Pass
# X963KDF : (KAT_KDF) : Pass
# X942KDF : (KAT_KDF) : Pass
# HASH : (DRBG) : Pass
# CTR : (DRBG) : Pass
# HMAC : (DRBG) : Pass
# DH : (KAT_KA) : Pass
# ECDH : (KAT_KA) : Pass
# RSA_Encrypt : (KAT_AsymmetricCipher) : Pass
# RSA_Decrypt : (KAT_AsymmetricCipher) : Pass
# RSA_Decrypt : (KAT_AsymmetricCipher) : Pass
# 
# On Demand Failure
# HMAC : (Module_Integrity) : Fail
# HMAC : (Install_Integrity) : Fail
# SHA1 : (KAT_Digest) : Fail
# SHA2 : (KAT_Digest) : Fail
# SHA3 : (KAT_Digest) : Fail
# TDES : (KAT_Cipher) : Fail
# AES_GCM : (KAT_Cipher) : Fail
# AES_ECB_Decrypt : (KAT_Cipher) : Fail
# RSA : (KAT_Signature) : Fail
# ECDSA : (PCT_Signature) : Fail
# ECDSA : (PCT_Signature) : Fail
# DSA : (PCT_Signature) : Fail
# TLS13_KDF_EXTRACT : (KAT_KDF) : Fail
# TLS13_KDF_EXPAND : (KAT_KDF) : Fail
# TLS12_PRF : (KAT_KDF) : Fail
# PBKDF2 : (KAT_KDF) : Fail
# SSHKDF : (KAT_KDF) : Fail
# KBKDF : (KAT_KDF) : Fail
# HKDF : (KAT_KDF) : Fail
# SSKDF : (KAT_KDF) : Fail
# X963KDF : (KAT_KDF) : Fail
# X942KDF : (KAT_KDF) : Fail
# HASH : (DRBG) : Fail
# CTR : (DRBG) : Fail
# HMAC : (DRBG) : Fail
# DH : (KAT_KA) : Fail
# ECDH : (KAT_KA) : Fail
# RSA_Encrypt : (KAT_AsymmetricCipher) : Fail
# RSA_Decrypt : (KAT_AsymmetricCipher) : Fail
# RSA_Decrypt : (KAT_AsymmetricCipher) : Fail
ok 1 - test_provider_status
```

**Reply by shuishuiwawa** [reply]
Have a test on both version 3.0.9 and the latest master code. Same result: the failed test cases after module integrity test did not run. I expect them to run as I mentioned above. I did the exact same things for various versions.
Steps:
1. Clone the repo: `git clone -b openssl-3.0.9 --single-branch https://github.com/openssl/openssl.git openssl309`
2. `cd openssl309 && ./config -d enable-fips shared --prefix=/opt/ossl309 && make && make install && make install_fips`
3. `LD_LIBRARY_PATH=/opt/ossl309/lib64 OPENSSL_CONF=/opt/ossl309/ssl/openssl-fips.cnf /opt/ossl309/bin/openssl fipsinstall -module /opt/ossl309/lib64/ossl-modules/fips.so -out /opt/ossl309/ssl/fipsmodule.cnf`
4. `cd /opt/ossl309/ssl && cp openssl.cnf openssl-fips.cnf`
5. `sed -i "/# .include fipsmodule.cnf/c\\\n.include /opt/ossl309/ssl/fipsmodule.cnf" openssl-fips.cnf && sed -i "/# fips = fips_sect/c\\\nfips = fips_sect\n" openssl-fips.cnf`
6. `make build_tests`
7. `test/provider_status_test -config /opt/ossl309/ssl/openssl-fips.cnf -provider_name fips`

Expected result: see the output above
Actual result: see the output in issue description

Also attach the output of `openssl version -a` and `openssl list -providers` 
![image](https://github.com/openssl/openssl/assets/10859404/d6e80ca0-2ec3-4f41-8531-c51b387c3bc0)
![image](https://github.com/openssl/openssl/assets/10859404/b2432442-059c-49de-b091-4f752703d969)

**Comment by shuishuiwawa** [comment, marked as answer]
Close it. I think the current behavior is by design. And for FIPS failure test cases, we should fail one at one time (then the result is failing to load the security module) and reboot/restart system/app to fail another one, repeating until all tests can fail. The current "Corruption" mechanism already supports failing one test case at one time. The only "issue" is that it  also goes through all the test cases instead of exiting the app on failure. I think it's also by design by checking the comments in self_test_kats.c.
